package connector;

option java_package = "jp.gaje.analog3.connector";
option java_outer_classname = "ConnectorMessages";

message Request {
  enum Command {
    DESCRIBE = 0;
    SET_ATTRIBUTE = 1;
    UNSET_ATTRIBUTE = 2;
    ADD_SUBCOMPONENT = 3;
    REMOVE_SUBCOMPONENT = 4;
  }

  required Command command = 1;
  repeated string path = 2;
  repeated uint32 id = 3;
  optional Attribute attribute = 4;
  optional Component component = 5;
}

message Reply {
  enum Status {
    SUCCESS = 0;
    ERROR = 1;
  }

  required Status status = 1;
  repeated Component component = 2;
}

message Component {
  required string name = 1;
  repeated Attribute attribute = 2;
  repeated Component sub_component = 3;
  optional uint32 id = 4;
}

message Attribute {
  required string name = 1;
  optional Value value = 2;
}

message Value {
  optional string svalue = 1;
  optional int32 ivalue = 2;
  optional double dvalue = 3;
  repeated Value avalue = 4;
}
